### ISSUE TITLE
`det` will return wrong gradient for `1x1` matrix with 0 value.

### ISSUE THREAD

**Issue by upupyy** [pytorch/pytorch#80761]
### 🐛 Describe the bug

`det` will return wrong gradient for `1x1` matrix with 0 value.

```python
import torch

input = torch.tensor([[0.]], dtype=torch.float64, requires_grad=True)
torch.det(input).backward()
print(input.grad)
# tensor([[0.]], dtype=torch.float64)
```

The correct gradient should be 1. Instead, when the value isn't zero, it will return the correct gradient.

```python
import torch

input = torch.tensor([[0.1]], dtype=torch.float64, requires_grad=True)
torch.det(input).backward()
print(input.grad)
# tensor([[1.]], dtype=torch.float64)
```

### Versions

pytorch: 1.11.0

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by Lezcano** [comment]
Det around singular matrices is tricky to compute. Its formula should be better in master since yesterday though.
